frontend/src/metabase/visualizations/visualizations/PivotTable/RowToggleIcon.jsx

```diff
@@ -0,0 +1,81 @@
+/* eslint-disable react/prop-types */
+import React from "react";
+import { updateIn } from "icepick";
+import _ from "underscore";
+
+import Icon from "metabase/components/Icon";
+
+import { COLLAPSED_ROWS_SETTING } from "metabase/lib/data_grid";
+
+import { RowToggleIconRoot } from "./PivotTable.styled";
+
+export function RowToggleIcon({
+  value,
+  settings,
+  updateSettings,
+  hideUnlessCollapsed,
+  rowIndex,
+  hasCustomColors,
+  isNightMode,
+}) {
+  if (value == null) {
+    return null;
+  }
+  const setting = settings[COLLAPSED_ROWS_SETTING];
+  const ref = JSON.stringify(value);
+  const isColumn = !Array.isArray(value);
+  const columnRef = isColumn ? null : JSON.stringify(value.length);
+  const settingValue = setting.value || [];
+  const isColumnCollapsed = !isColumn && settingValue.includes(columnRef);
+  const isCollapsed = settingValue.includes(ref) || isColumnCollapsed;
+  if (hideUnlessCollapsed && !isCollapsed) {
+    // subtotal rows shouldn't have an icon unless the section is collapsed
+    return null;
+  }
+
+  // The giant nested ternary below picks the right function to toggle the current button.
+  // That depends on whether we're a row or column header and whether we're open or closed.
+  const toggle =
+    isColumn && !isCollapsed // click on open column
+      ? settingValue =>
```

Comment:
I wouldn't mind at all if we later converted this cascade of ternaries into if/else blocks.